simulation/sub8_simulation/sub8_sim_tools/physics/vehicle.py

```diff
@@ -149,4 +255,42 @@ def step(self, dt):
         self.apply_buoyancy_force()
         self.apply_damping_force()
         # self.apply_damping_torque()
-        self.publish_pose()
\ No newline at end of file
+        self.publish_pose()
+        self.publish_dvl()
+        self.publish_imu()
+
+        # Sets the position of the DVL sensor rays to point below the current position of the sub
+        self.dvl_rays[0].set(self.body.getRelPointPos(self.dvl_position), self.body.vectorToWorld(np.array([0.866, .5, -1])))
+        self.dvl_rays[1].set(self.body.getRelPointPos(self.dvl_position), self.body.vectorToWorld(np.array([0.866, -.5, -1])))
+        self.dvl_rays[2].set(self.body.getRelPointPos(self.dvl_position), self.body.vectorToWorld(np.array([-0.866, .5, -1])))
+        self.dvl_rays[3].set(self.body.getRelPointPos(self.dvl_position), self.body.vectorToWorld(np.array([-0.866, -.5, -1])))
+
+    def get_dvl_range(self):
+        """Returns the range of the ray, otherwise 0 represents contact with another object
+        For each dvl sensor ray object, collisions with the floor are detected, and parameters from the contact.getContactGeomParams() method 
+        are used to calculate the range of the dvl sensor to the floor - returns 0 if unable to get a measurement (unable to connect with the floor)"""
+        for ray in self.dvl_rays:
+            for contact in ode.collide(ray, self.space):
+                pos, normal, depth, geom1, geom2 = contact.getContactGeomParams()
+
+                assert geom1 is ray, geom1
+                if (geom2 == self.geom):
+                    continue
+                print (self.body.getRelPointPos(self.dvl_position) - np.array(pos))[2]
+
+        return 0
+
+
+    def get_dvl_vel(self):
+        """Returns array of 4 components of the vehicle's velocities based off of the dvl ray vectors"""
+        dvl_ray_1 = np.array([0.866, -.5, -1])
+        dvl_ray_2 = np.array([0.866, .5, -1])
+        dvl_ray_3 = np.array([-0.866, .5, -1])
+        dvl_ray_4 = np.array([-0.866, -.5, -1])
+
+        vel_sub_world = self.body.vectorFromWorld(self.body.getRelPointVel(self.dvl_position))
```

Comment:
This is definitely a body-frame quantity, so it's probably better named `vel_sub_body`. It's a bit more subtle than that, though, since it's not the velocity of the sub, but the velocity of the DVL ... so `vel_dvl_body` is probably better yet.